src/wiktextract/extractor/simple/page_utils.py

```diff
@@ -0,0 +1,50 @@
+from typing import Union
+
+from wikitextprocessor import LevelNode, WikiNode
+from wiktextract.page import clean_node
+from wiktextract.wxr_context import WiktextractContext
+
+from .etymology import process_etym
+from .models import WordEntry
+from .pronunciation import process_pron
+
+Node = Union[str, WikiNode]
+
+def separate_subsections(node: WikiNode) -> tuple[list[Node], list[WikiNode]]:
+    i = 0  # needed if for loop doesn't run
+    for i, child in enumerate(node.children):
+        if isinstance(child, LevelNode):
+            break
+    else:  # when ending for loop, if there was no break:
+        i += 1
+    # We are pretty much guaranteed that the second half of children is only
+    # level kinds
+    return node.children[:i], node.children[i:]  # type: ignore[return-value]
+
+
+def recurse_base_data_sections(
+    wxr: WiktextractContext, node: WikiNode, target_data: WordEntry
+) -> None:
+    # Parse a section recursively; a heading node with contents and possible
+    # subsections (always following the non-subsection content)
+    content, subsections = separate_subsections(node)
```

Comment:
Why not use the parsed node hierarchy directly:

```python
title_text = clean_node(wxr, None, level_node.largs).lower()
if title_text.startswith(("etymology", "word parts")):
    pass
elif title_text.startswith("..."):
    pass
for next_level in level_node.find_child(LEVEL_KIND_FLAGS):
    recurse_base_data_sections(wxr, next_level, target_data)

```

Reply:
That's remnants of when "contents" was/is useful as a clump. If you want to run it through clean_node() whole, then you can just do `clean_node(..., contents, ...)`. I'll clean it up.

Reply:
Looking at this, I will keep things as is. There is nothing harmful here, and as the processing functions are they take a whole `contents` block as is, which is needed because we want a contents-wide context.

Reply:
Usually I'd like to pass the level node to each section's extract function, because I could then use `WikiNode.find_child()` on this node. If I want to convert the text of this section but exclude subsection I could pass `WikiNode.invert_find_child(LEVEL_KIND_FLAGS)` to `clean_node()`.

Reply:
That sounds good. I've removed the preprocessing (except for removing `text block start` template stuff), so now the main page loop is going through each level in order.

One problem with this approach is that *sometimes* you have a situation like this:

```
== Noun ==
...
=== Pronunciation ===
...
=== Word parts ===
...
```

at the end of an article, so the pronunciation and word parts stuff is meant to go with the Noun, instead of preceding it (as is the usual way). This is why I ended up with the preprocessing stuff earlier.

Reply:
I have the similar problem in zh edition, I use a recursive [`parse_section()`](https://github.com/tatuylonen/wiktextract/blob/3218b933c35638294b6886a4bdb0d94e2bcc3788/src/wiktextract/extractor/zh/page.py#L67) function processing any level node at any place. It adds the extracted data to the last `WordEntry` if we already have some POS section data or adds to `base_data`. Some sections probably do more than that, but I think it's the general idea.

Reply:
I solved this by changing from iterators to lists and also parsing heading titles before-hand. That way, each time parsing a pron/etym section, you can check if any of the following titles are POS headings and if not, use the last POS data instead of putting stuff in `base_data`.